# qutip - installation checking

In [4]:
from qutip import *

In [5]:
about()


QuTiP: Quantum Toolbox in Python
Copyright (c) QuTiP team 2011 and later.
Original developers: R. J. Johansson & P. D. Nation.
Previous lead developers: Chris Granade & A. Grimsmo.
Current admin team: Alexander Pitchford, Paul D. Nation, Nathan Shammah, Shahnawaz Ahmed, Neill Lambert, and Eric Giguère.
Project Manager: Franco Nori.
Currently developed through wide collaboration. See https://github.com/qutip for details.

QuTiP Version:      4.5.0
Numpy Version:      1.18.5
Scipy Version:      1.5.0
Cython Version:     0.29.20
Matplotlib Version: 3.2.2
Python Version:     3.7.7
Number of CPUs:     4
BLAS Info:          INTEL MKL
OPENMP Installed:   False
INTEL MKL Ext:      True
Platform Info:      Windows (AMD64)
Installation path:  C:\Users\NAVS\anaconda3\envs\qutip-env\lib\site-packages\qutip
Please cite QuTiP in your publication.
For your convenience a bibtex reference can be easily generated using `qutip.cite()`


In [6]:
import qutip.testing as qt

In [8]:
qt.run()


QuTiP: Quantum Toolbox in Python
Copyright (c) QuTiP team 2011 and later.
Original developers: R. J. Johansson & P. D. Nation.
Previous lead developers: Chris Granade & A. Grimsmo.
Current admin team: Alexander Pitchford, Paul D. Nation, Nathan Shammah, Shahnawaz Ahmed, Neill Lambert, and Eric Giguère.
Project Manager: Franco Nori.
Currently developed through wide collaboration. See https://github.com/qutip for details.

QuTiP Version:      4.5.0
Numpy Version:      1.18.5
Scipy Version:      1.5.0
Cython Version:     0.29.20
Matplotlib Version: 3.2.2
Python Version:     3.7.7
Number of CPUs:     4
BLAS Info:          INTEL MKL
OPENMP Installed:   False
INTEL MKL Ext:      True
Platform Info:      Windows (AMD64)
Installation path:  C:\Users\NAVS\anaconda3\envs\qutip-env\lib\site-packages\qutip
Please cite QuTiP in your publication.
For your convenience a bibtex reference can be easily generated using `qutip.cite()`
============================= test session starts ===================

# Pytest output
-----------------------------------

QuTiP: Quantum Toolbox in Python
================================
Copyright (c) QuTiP team 2011 and later.
Original developers: R. J. Johansson & P. D. Nation.
Previous lead developers: Chris Granade & A. Grimsmo.
Current admin team: Alexander Pitchford, Paul D. Nation, Nathan Shammah, Shahnawaz Ahmed, Neill Lambert, and Eric Giguère.
Project Manager: Franco Nori.
Currently developed through wide collaboration. See https://github.com/qutip for details.

QuTiP Version:      4.5.0
Numpy Version:      1.18.5
Scipy Version:      1.5.0
Cython Version:     0.29.20
Matplotlib Version: 3.2.2
Python Version:     3.7.7
Number of CPUs:     4
BLAS Info:          INTEL MKL
OPENMP Installed:   False
INTEL MKL Ext:      True
Platform Info:      Windows (AMD64)
Installation path:  C:\Users\NAVS\anaconda3\envs\qutip-env\lib\site-packages\qutip
==============================================================================
Please cite QuTiP in your publication.
==============================================================================
For your convenience a bibtex reference can be easily generated using `qutip.cite()`
============================= test session starts =============================
platform win32 -- Python 3.7.7, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- C:\Users\NAVS\anaconda3\envs\qutip-env\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\NAVS\1---conda-start\2-qutip
collecting ... collected 703 items / 61 deselected / 642 selected

tests/test_basis_transformation.py::test_transformation_to_eigenbasis_is_reversible[2-real_hermitian] PASSED [  0%]
tests/test_basis_transformation.py::test_transformation_to_eigenbasis_is_reversible[2-imaginary_hermitian] PASSED [  0%]
tests/test_basis_transformation.py::test_transformation_to_eigenbasis_is_reversible[2-complex_hermitian] PASSED [  0%]
tests/test_basis_transformation.py::test_transformation_to_eigenbasis_is_reversible[10-real_hermitian] PASSED [  0%]
tests/test_basis_transformation.py::test_transformation_to_eigenbasis_is_reversible[10-imaginary_hermitian] PASSED [  0%]
tests/test_basis_transformation.py::test_transformation_to_eigenbasis_is_reversible[10-complex_hermitian] PASSED [  0%]
tests/test_basis_transformation.py::test_ket_and_dm_transformations_equivalent[4] PASSED [  1%]
tests/test_basis_transformation.py::test_eigenbasis_transformation_makes_diagonal_operator PASSED [  1%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-2-rand_herm] PASSED [  1%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-2-rand_ket] PASSED [  1%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-2-rand_bra] FAILED [  1%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-10-rand_herm] PASSED [  1%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-10-rand_ket] PASSED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-10-rand_bra] FAILED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-2-rand_herm] PASSED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-2-rand_ket] PASSED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-2-rand_bra] FAILED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-10-rand_herm] PASSED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-10-rand_ket] PASSED [  2%]
tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-10-rand_bra] FAILED [  3%]
tests/test_brmesolve.py::test_simple_qubit_system[me_c_ops0-brme_c_ops0-brme_a_ops0] PASSED [  3%]
tests/test_brmesolve.py::test_simple_qubit_system[me_c_ops1-brme_c_ops1-brme_a_ops1] PASSED [  3%]
tests/test_brmesolve.py::test_simple_qubit_system[me_c_ops2-brme_c_ops2-brme_a_ops2] PASSED [  3%]
tests/test_brmesolve.py::test_harmonic_oscillator[0] PASSED              [  3%]
tests/test_brmesolve.py::test_harmonic_oscillator[0.15] PASSED           [  3%]
tests/test_brmesolve.py::test_jaynes_cummings_zero_temperature PASSED    [  4%]
tests/test_brmesolve.py::test_pull_572_error PASSED                      [  4%]
tests/test_brmesolve.py::test_solver_accepts_list_hamiltonian PASSED     [  4%]
tests/test_brmesolve_td.py::test_time_dependent_spline_in_c_ops PASSED   [  4%]
tests/test_brmesolve_td.py::test_split_operators_maintain_answer[_2_tuple_split] PASSED [  4%]
tests/test_brtools.py::test_zheevr PASSED                                [  4%]
tests/test_brtools.py::test_dense_operator_to_eigbasis[random Hermitian] PASSED [  4%]
tests/test_brtools.py::test_dense_operator_to_eigbasis[annihilation] PASSED [  5%]
tests/test_brtools.py::test_vec_to_eigbasis PASSED                       [  5%]
tests/test_brtools.py::test_eigvec_to_fockbasis PASSED                   [  5%]
tests/test_brtools.py::test_vector_roundtrip PASSED                      [  5%]
tests/test_brtools.py::test_diag_liou_mult PASSED                        [  5%]
tests/test_brtools.py::test_cop_super_mult PASSED                        [  5%]
tests/test_brtools.py::test_br_term_mult[secular] PASSED                 [  6%]
tests/test_brtools.py::test_br_term_mult[non-secular] PASSED             [  6%]
tests/test_cavityqed.py::Testcqed::test_DispersiveCavityQED_ISWAP PASSED [  6%]
tests/test_cavityqed.py::Testcqed::test_DispersiveCavityQED_combination PASSED [  6%]
tests/test_cavityqed.py::Testcqed::test_analytical_evo PASSED            [  6%]
tests/test_cavityqed.py::Testcqed::test_numerical_evo PASSED             [  6%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_01_1_unitary_hadamard PASSED [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_01_2_unitary_hadamard_no_stats PASSED [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_01_3_unitary_hadamard_tau PASSED [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_01_4_unitary_hadamard_qobj PASSED [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_01_5_unitary_hadamard_oo PASSED [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_01_6_unitary_hadamard_grad PASSED [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_02_1_qft PASSED   [  7%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_02_2_qft_bounds PASSED [  8%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_03_dumping PASSED [  8%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_04_unitarity PASSED [  8%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_05_1_state_to_state PASSED [  8%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_05_2_state_to_state_qobj PASSED [  8%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_06_lindbladian PASSED [  8%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_07_symplectic PASSED [  9%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_08_crab PASSED    [  9%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_09_load_params PASSED [  9%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_10_init_pulse_params PASSED [  9%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_11_time_dependent_drift PASSED [  9%]
tests/test_control_pulseoptim.py::TestPulseOptim::test_12_time_dependent_ctrls PASSED [  9%]
tests/test_correlation.py::test_compare_solvers_coherent_state_legacy PASSED [  9%]
tests/test_correlation.py::test_compare_solvers_coherent_state_mees PASSED [ 10%]
tests/test_correlation.py::test_compare_solvers_steadystate_legacy PASSED [ 10%]
tests/test_correlation.py::test_compare_solvers_steadystate PASSED       [ 10%]
tests/test_correlation.py::test_spectrum_espi_legacy PASSED              [ 10%]
tests/test_correlation.py::test_spectrum_esfft PASSED                    [ 10%]
tests/test_correlation.py::test_spectrum_espi PASSED                     [ 10%]
tests/test_correlation.py::test_H_str_list_td_corr PASSED                [ 11%]
tests/test_correlation.py::test_H_np_list_td_corr PASSED                 [ 11%]
tests/test_correlation.py::test_H_fn_list_td_corr PASSED                 [ 11%]
tests/test_correlation.py::test_H_fn_td_corr PASSED                      [ 11%]
tests/test_correlation.py::test_c_ops_fn_list_td_corr PASSED             [ 11%]
tests/test_correlation.py::test_np_list_td_corr PASSED                   [ 11%]
tests/test_correlation.py::test_fn_list_td_corr PASSED                   [ 11%]
tests/test_correlation.py::test_fn_list_td_corr2 PASSED                  [ 12%]
tests/test_countstat.py::test_dqd_current FAILED                         [ 12%]
tests/test_cy_structs.py::test_coo2csr_struct PASSED                     [ 12%]
tests/test_cy_structs.py::test_indices_sort PASSED                       [ 12%]
tests/test_cy_structs.py::test_coo2csr_inplace_nosort PASSED             [ 12%]
tests/test_cy_structs.py::test_coo2csr_inplace_sort PASSED               [ 12%]
tests/test_cy_structs.py::test_csr2coo PASSED                            [ 13%]
tests/test_dimensions.py::test_flatten PASSED                            [ 13%]
tests/test_dimensions.py::test_enumerate_flat PASSED                     [ 13%]
tests/test_dimensions.py::test_deep_remove PASSED                        [ 13%]
tests/test_dimensions.py::test_unflatten PASSED                          [ 13%]
tests/test_dimensions.py::test_dims_idxs_to_tensor_idxs PASSED           [ 13%]
tests/test_dimensions.py::test_dims_to_tensor_shape PASSED               [ 14%]
tests/test_dimensions.py::test_type_from_dims PASSED                     [ 14%]
tests/test_dimensions.py::test_collapse PASSED                           [ 14%]
tests/test_eigenstates.py::test_diagHamiltonian1 PASSED                  [ 14%]
tests/test_eigenstates.py::test_diagHamiltonian2 PASSED                  [ 14%]
tests/test_enr_state_operator.py::test_enr_destory_full PASSED           [ 14%]
tests/test_enr_state_operator.py::test_enr_destory_single PASSED         [ 14%]
tests/test_enr_state_operator.py::test_enr_destory_double PASSED         [ 15%]
tests/test_enr_state_operator.py::test_enr_fock_state PASSED             [ 15%]
tests/test_enr_state_operator.py::test_enr_identity PASSED               [ 15%]
tests/test_enr_state_operator.py::test_enr_thermal_dm1 PASSED            [ 15%]
tests/test_enr_state_operator.py::test_enr_thermal_dm2 PASSED            [ 15%]
tests/test_entropy.py::test_EntropyVN PASSED                             [ 15%]
tests/test_entropy.py::test_EntropyLinear PASSED                         [ 16%]
tests/test_entropy.py::test_EntropyConcurrence PASSED                    [ 16%]
tests/test_entropy.py::test_EntropyMutual PASSED                         [ 16%]
tests/test_entropy.py::test_EntropyConditional PASSED                    [ 16%]
tests/test_entropy.py::test_EntanglingPower PASSED                       [ 16%]
tests/test_expect.py::TestExpect::testOperatorKet PASSED                 [ 16%]
tests/test_expect.py::TestExpect::testOperatorKetRand PASSED             [ 16%]
tests/test_expect.py::TestExpect::testOperatorDensityMatrix PASSED       [ 17%]
tests/test_expect.py::TestExpect::testOperatorStateList PASSED           [ 17%]
tests/test_expect.py::TestExpect::testOperatorListState PASSED           [ 17%]
tests/test_expect.py::TestExpect::testOperatorListStateList PASSED       [ 17%]
tests/test_expect.py::TestExpect::testExpectSolverCompatibility PASSED   [ 17%]
tests/test_fastsparse.py::test_fast_sparse_basic FAILED                  [ 17%]
tests/test_fastsparse.py::test_fast_sparse_trans PASSED                  [ 18%]
tests/test_fastsparse.py::test_fast_sparse_adjoint PASSED                [ 18%]
tests/test_fileio.py::TestFileIO::testRWRealDefault PASSED               [ 18%]
tests/test_fileio.py::TestFileIO::testRWRealDecimal PASSED               [ 18%]
tests/test_fileio.py::TestFileIO::testRWRealExp PASSED                   [ 18%]
tests/test_fileio.py::TestFileIO::testRWComplexDefault PASSED            [ 18%]
tests/test_fileio.py::TestFileIO::testRWComplexDecimal PASSED            [ 19%]
tests/test_fileio.py::TestFileIO::testRWComplexExp PASSED                [ 19%]
tests/test_fileio.py::TestFileIO::testRWSeparatorDetection PASSED        [ 19%]
tests/test_fileio.py::TestFileIO::testqsaveqload PASSED                  [ 19%]
tests/test_floquet.py::TestFloquet::testFloquetUnitary PASSED            [ 19%]
tests/test_gates.py::TestGates::testSwapGate PASSED                      [ 19%]
tests/test_gates.py::TestGates::test_clifford_group_len PASSED           [ 19%]
tests/test_gates.py::TestGates::test_are_cliffords PASSED                [ 20%]
tests/test_gates.py::TestGates::testExpandGate1toN PASSED                [ 20%]
tests/test_gates.py::TestGates::testExpandGate2toNSwap PASSED            [ 20%]
tests/test_gates.py::TestGates::testExpandGate2toN PASSED                [ 20%]
tests/test_gates.py::TestGates::testExpandGate3toN_permutation PASSED    [ 20%]
tests/test_gates.py::TestGates::testExpandGate3toN PASSED                [ 20%]
tests/test_gates.py::TestGates::test_expand_operator PASSED              [ 21%]
tests/test_gates.py::TestGates::test_molmer_sorensen PASSED              [ 21%]
tests/test_gates.py::TestGates::test_qrot PASSED                         [ 21%]
tests/test_graph.py::test_graph_degree PASSED                            [ 21%]
tests/test_graph.py::test_graph_bfs PASSED                               [ 21%]
tests/test_graph.py::test_graph_rcm_simple PASSED                        [ 21%]
tests/test_graph.py::test_graph_rcm_boost PASSED                         [ 21%]
tests/test_graph.py::test_graph_rcm_qutip PASSED                         [ 22%]
tests/test_graph.py::test_graph_maximum_bipartite_matching PASSED        [ 22%]
tests/test_graph.py::test_graph_weighted_matching PASSED                 [ 22%]
tests/test_graph.py::test_column_permutation PASSED                      [ 22%]
tests/test_heom_solver.py::TestHSolver::test_pure_dephasing PASSED       [ 22%]
tests/test_interpolate.py::testInterpolate1 PASSED                       [ 22%]
tests/test_interpolate.py::testInterpolate2 PASSED                       [ 23%]
tests/test_interpolate.py::testInterpolate3 PASSED                       [ 23%]
tests/test_interpolate.py::testInterpolate4 PASSED                       [ 23%]
tests/test_lattice.py::TestLattice::test_hamiltonian PASSED              [ 23%]
tests/test_lattice.py::TestLattice::test_cell_structures PASSED          [ 23%]
tests/test_lattice.py::TestLattice::test_basis PASSED                    [ 23%]
tests/test_lattice.py::TestLattice::test_distribute_operator PASSED      [ 23%]
tests/test_lattice.py::TestLattice::test_operator_at_cells PASSED        [ 24%]
tests/test_lattice.py::TestLattice::test_operator_between_cells PASSED   [ 24%]
tests/test_lattice.py::TestLattice::test_x PASSED                        [ 24%]
tests/test_lattice.py::TestLattice::test_k PASSED                        [ 24%]
tests/test_lattice.py::TestLattice::test_get_dispersion PASSED           [ 24%]
tests/test_lattice.py::TestLattice::test_cell_periodic_parts PASSED      [ 24%]
tests/test_lattice.py::TestLattice::test_bulk_Hamiltonians PASSED        [ 25%]
tests/test_lattice.py::TestLattice::test_bloch_wave_functions FAILED     [ 25%]
tests/test_lattice.py::TestLattice::test_CROW FAILED                     [ 25%]
tests/test_lattice.py::TestLattice::test_SSH PASSED                      [ 25%]
tests/test_mcsolve.py::test_MCNoCollExpt PASSED                          [ 25%]
tests/test_mcsolve.py::test_MCNoCollStates PASSED                        [ 25%]
tests/test_mcsolve.py::test_MCNoCollExpectStates PASSED                  [ 26%]
tests/test_mcsolve.py::test_MCNoCollFuncExpt PASSED                      [ 26%]
tests/test_mcsolve.py::test_MCNoCollFuncStates PASSED                    [ 26%]
tests/test_mcsolve.py::test_MCCollapseTimesOperators PASSED              [ 26%]
tests/test_mcsolve.py::test_MCTDFunc PASSED                              [ 26%]
tests/test_mcsolve.py::test_MCTDStr PASSED                               [ 26%]
tests/test_mcsolve.py::test_mc_dtypes1 PASSED                            [ 26%]
tests/test_mcsolve.py::test_mc_dtypes2 PASSED                            [ 27%]
tests/test_mcsolve.py::test_mc_dyn_args PASSED                           [ 27%]
tests/test_mcsolve.py::test_mc_functd_sum PASSED                         [ 27%]
tests/test_mesolve.py::TestJCModelEvolution::testQubitDynamics1 PASSED   [ 27%]
tests/test_mesolve.py::TestJCModelEvolution::testQubitDynamics2 PASSED   [ 27%]
tests/test_mesolve.py::TestJCModelEvolution::testCase1 PASSED            [ 27%]
tests/test_mesolve.py::TestJCModelEvolution::testCase2 PASSED            [ 28%]
tests/test_mesolve.py::TestJCModelEvolution::testCase3 PASSED            [ 28%]
tests/test_mesolve.py::TestMESolverConstDecay::testMEDecay PASSED        [ 28%]
tests/test_mesolve.py::TestMESolverConstDecay::testMEDecaySingleCollapse PASSED [ 28%]
tests/test_mesolve.py::TestMESolverConstDecay::testMEDecayAsFuncList PASSED [ 28%]
tests/test_mesolve.py::TestMESolverConstDecay::testMEDecayAsStrList PASSED [ 28%]
tests/test_mesolve.py::TestMESolveTDDecay::testMETDDecayAsFuncList PASSED [ 28%]
tests/test_mesolve.py::TestMESolveTDDecay::testMETDDecayAsPartFuncList PASSED [ 29%]
tests/test_mesolve.py::TestMESolveTDDecay::testMETDDecayAsStrList PASSED [ 29%]
tests/test_mesolve.py::TestMESolveTDDecay::testMETDDecayAsArray PASSED   [ 29%]
tests/test_mesolve.py::TestMESolveTDDecay::testMETDDecayAsFunc PASSED    [ 29%]
tests/test_mesolve.py::TestMESolveSuperInit::testSuperJC PASSED          [ 29%]
tests/test_mesolve.py::TestMESolveSuperInit::testMETDDecayAsFuncList PASSED [ 29%]
tests/test_mesolve.py::TestMESolveSuperInit::testMETDDecayAsPartFuncList PASSED [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::testMETDDecayAsStrList PASSED [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::testMETDDecayAsArray PASSED [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::testMETDDecayAsFunc PASSED  [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::test_me_interp1 PASSED      [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::test_me_interp2 PASSED      [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::test_me_interp3 PASSED      [ 30%]
tests/test_mesolve.py::TestMESolveSuperInit::test_me_interp4 PASSED      [ 31%]
tests/test_mesolve.py::TestMESolverMisc::testMEFinalState PASSED         [ 31%]
tests/test_mesolve.py::TestMESolverMisc::testSEFinalState PASSED         [ 31%]
tests/test_mesolve.py::TestMESolveStepFuncCoeff::test_py_coeff PASSED    [ 31%]
tests/test_mesolve.py::TestMESolveStepFuncCoeff::test_array_cte_coeff PASSED [ 31%]
tests/test_mesolve.py::TestMESolveStepFuncCoeff::test_array_t_coeff PASSED [ 31%]
tests/test_mesolve.py::TestMESolveStepFuncCoeff::test_array_str_coeff PASSED [ 32%]
tests/test_mesolve.py::TestMESolveStepFuncCoeff::test_array_str_py_coeff PASSED [ 32%]
tests/test_mesolve.py::TestMESolveStepFuncCoeff::test_dynamic_args PASSED [ 32%]
tests/test_metrics.py::test_fid_trdist_limits PASSED                     [ 32%]
tests/test_metrics.py::test_fidelity1 PASSED                             [ 32%]
tests/test_metrics.py::test_fidelity2 PASSED                             [ 32%]
tests/test_metrics.py::test_fidelity_max PASSED                          [ 33%]
tests/test_metrics.py::test_fidelity_bounded_purepure PASSED             [ 33%]
tests/test_metrics.py::test_fidelity_bounded_puremixed PASSED            [ 33%]
tests/test_metrics.py::test_fidelity_bounded_mixedmixed PASSED           [ 33%]
tests/test_metrics.py::test_fidelity_known_cases PASSED                  [ 33%]
tests/test_metrics.py::test_fidelity_overlap PASSED                      [ 33%]
tests/test_metrics.py::test_tracedist1 PASSED                            [ 33%]
tests/test_metrics.py::test_tracedist2 PASSED                            [ 34%]
tests/test_metrics.py::test_tracedist3 PASSED                            [ 34%]
tests/test_metrics.py::test_hellinger_corner PASSED                      [ 34%]
tests/test_metrics.py::test_hellinger_pure PASSED                        [ 34%]
tests/test_metrics.py::test_hellinger_inequality PASSED                  [ 34%]
tests/test_metrics.py::test_hellinger_monotonicity PASSED                [ 34%]
tests/test_metrics.py::test_average_gate_fidelity PASSED                 [ 35%]
tests/test_metrics.py::test_average_gate_fidelity_target PASSED          [ 35%]
tests/test_metrics.py::test_hilbert_dist PASSED                          [ 35%]
tests/test_metrics.py::test_unitarity_known XFAIL                        [ 35%]
tests/test_metrics.py::test_unitarity_bounded XFAIL                      [ 35%]
tests/test_metrics.py::test_dnorm_bounded SKIPPED                        [ 35%]
tests/test_metrics.py::test_dnorm_qubit_known_cases SKIPPED              [ 35%]
tests/test_metrics.py::test_dnorm_qubit_scalar SKIPPED                   [ 36%]
tests/test_metrics.py::test_dnorm_qubit_triangle SKIPPED                 [ 36%]
tests/test_metrics.py::test_dnorm_force_solve SKIPPED                    [ 36%]
tests/test_metrics.py::test_dnorm_cptp SKIPPED                           [ 36%]
tests/test_mkl.py::test_mkl_spsolve1 PASSED                              [ 36%]
tests/test_mkl.py::test_mklspsolve2 PASSED                               [ 36%]
tests/test_mkl.py::test_mkl_spsolve3 PASSED                              [ 37%]
tests/test_mkl.py::test_mkl_spsolve4 PASSED                              [ 37%]
tests/test_mkl.py::test_mkl_spsolve5 PASSED                              [ 37%]
tests/test_mkl.py::test_mkl_spsolve6 PASSED                              [ 37%]
tests/test_mkl.py::test_mkl_spsolve7 PASSED                              [ 37%]
tests/test_mkl.py::test_mkl_spsolve8 PASSED                              [ 37%]
tests/test_mkl.py::test_mkl_spsolve9 PASSED                              [ 38%]
tests/test_mkl.py::test_mkl_spsolve10 PASSED                             [ 38%]
tests/test_openmp.py::test_openmp_spmv SKIPPED                           [ 38%]
tests/test_openmp.py::test_openmp_mesolve SKIPPED                        [ 38%]
tests/test_openmp.py::test_openmp_mesolve_td SKIPPED                     [ 38%]
tests/test_operators.py::test_jmat_12 PASSED                             [ 38%]
tests/test_operators.py::test_jmat_32 PASSED                             [ 38%]
tests/test_operators.py::test_jmat_42 PASSED                             [ 39%]
tests/test_operators.py::test_jmat_52 PASSED                             [ 39%]
tests/test_operators.py::test_destroy PASSED                             [ 39%]
tests/test_operators.py::test_create PASSED                              [ 39%]
tests/test_operators.py::test_simple_operator_creation[1-qzero-<lambda>] PASSED [ 39%]
tests/test_operators.py::test_simple_operator_creation[1-qeye-<lambda>] PASSED [ 39%]
tests/test_operators.py::test_simple_operator_creation[5-qzero-<lambda>] PASSED [ 40%]
tests/test_operators.py::test_simple_operator_creation[5-qeye-<lambda>] PASSED [ 40%]
tests/test_operators.py::test_simple_operator_creation[100-qzero-<lambda>] PASSED [ 40%]
tests/test_operators.py::test_simple_operator_creation[100-qeye-<lambda>] PASSED [ 40%]
tests/test_operators.py::test_implicit_tensor_creation[2-qzero] PASSED   [ 40%]
tests/test_operators.py::test_implicit_tensor_creation[2-qeye] PASSED    [ 40%]
tests/test_operators.py::test_implicit_tensor_creation[2-identity] PASSED [ 40%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions1-qzero] PASSED [ 41%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions1-qeye] PASSED [ 41%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions1-identity] PASSED [ 41%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions2-qzero] PASSED [ 41%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions2-qeye] PASSED [ 41%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions2-identity] PASSED [ 41%]
tests/test_operators.py::test_implicit_tensor_creation[1-qzero] PASSED   [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[1-qeye] PASSED    [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[1-identity] PASSED [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions4-qzero] PASSED [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions4-qeye] PASSED [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions4-identity] PASSED [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions5-qzero] PASSED [ 42%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions5-qeye] PASSED [ 43%]
tests/test_operators.py::test_implicit_tensor_creation[dimensions5-identity] PASSED [ 43%]
tests/test_operators.py::test_super_operator_creation[qzero] PASSED      [ 43%]
tests/test_operators.py::test_super_operator_creation[qeye] PASSED       [ 43%]
tests/test_operators.py::test_super_operator_creation[identity] PASSED   [ 43%]
tests/test_operators.py::test_num PASSED                                 [ 43%]
tests/test_operators.py::test_squeeze PASSED                             [ 44%]
tests/test_operators.py::test_displace PASSED                            [ 44%]
tests/test_operators.py::test_charge PASSED                              [ 44%]
tests/test_operators.py::test_tunneling PASSED                           [ 44%]
tests/test_optpulseprocessor.py::TestOptPulseProcessor::test_simple_hadamard PASSED [ 44%]
tests/test_optpulseprocessor.py::TestOptPulseProcessor::test_multi_qubits PASSED [ 44%]
tests/test_optpulseprocessor.py::TestOptPulseProcessor::test_multi_gates PASSED [ 45%]
tests/test_parallel.py::test_parfor1 PASSED                              [ 45%]
tests/test_parallel.py::test_parallel_map PASSED                         [ 45%]
tests/test_parallel.py::test_serial_map PASSED                           [ 45%]
tests/test_partial_transpose.py::test_partial_transpose_bipartite PASSED [ 45%]
tests/test_partial_transpose.py::test_partial_transpose_comparison PASSED [ 45%]
tests/test_partial_transpose.py::test_partial_transpose_randomized PASSED [ 45%]
tests/test_piqs.py::TestDicke::test_num_dicke_states PASSED              [ 46%]
tests/test_piqs.py::TestDicke::test_num_tls PASSED                       [ 46%]
tests/test_piqs.py::TestDicke::test_num_dicke_ladders PASSED             [ 46%]
tests/test_piqs.py::TestDicke::test_j_vals PASSED                        [ 46%]
tests/test_piqs.py::TestDicke::test_m_vals PASSED                        [ 46%]
tests/test_piqs.py::TestDicke::test_dicke_blocks PASSED                  [ 46%]
tests/test_piqs.py::TestDicke::test_dicke_blocks_full PASSED             [ 47%]
tests/test_piqs.py::TestDicke::test_dicke_function_trace PASSED          [ 47%]
tests/test_piqs.py::TestDicke::test_entropy_vn_dicke PASSED              [ 47%]
tests/test_piqs.py::TestDicke::test_purity_dicke PASSED                  [ 47%]
tests/test_piqs.py::TestDicke::test_get_index PASSED                     [ 47%]
tests/test_piqs.py::TestDicke::test_jmm1_dictionary PASSED               [ 47%]
tests/test_piqs.py::TestDicke::test_lindbladian PASSED                   [ 47%]
tests/test_piqs.py::TestDicke::test_gamma PASSED                         [ 48%]
tests/test_piqs.py::TestDicke::test_jspin PASSED                         [ 48%]
tests/test_piqs.py::TestDicke::test_j_min_ PASSED                        [ 48%]
tests/test_piqs.py::TestDicke::test_energy_degeneracy PASSED             [ 48%]
tests/test_piqs.py::TestDicke::test_state_degeneracy PASSED              [ 48%]
tests/test_piqs.py::TestDicke::test_m_degeneracy PASSED                  [ 48%]
tests/test_piqs.py::TestDicke::test_ap PASSED                            [ 49%]
tests/test_piqs.py::TestDicke::test_am PASSED                            [ 49%]
tests/test_piqs.py::TestDicke::test_spin_algebra PASSED                  [ 49%]
tests/test_piqs.py::TestDicke::test_collective_algebra PASSED            [ 49%]
tests/test_piqs.py::TestDicke::test_block_matrix PASSED                  [ 49%]
tests/test_piqs.py::TestDicke::test_dicke_basis PASSED                   [ 49%]
tests/test_piqs.py::TestDicke::test_dicke PASSED                         [ 50%]
tests/test_piqs.py::TestDicke::test_excited PASSED                       [ 50%]
tests/test_piqs.py::TestDicke::test_superradiant PASSED                  [ 50%]
tests/test_piqs.py::TestDicke::test_ghz PASSED                           [ 50%]
tests/test_piqs.py::TestDicke::test_ground PASSED                        [ 50%]
tests/test_piqs.py::TestDicke::test_identity_uncoupled PASSED            [ 50%]
tests/test_piqs.py::TestDicke::test_css PASSED                           [ 50%]
tests/test_piqs.py::TestDicke::test_collapse_uncoupled PASSED            [ 51%]
tests/test_piqs.py::TestDicke::test_get_blocks PASSED                    [ 51%]
tests/test_piqs.py::TestDicke::test_lindbladian_dims PASSED              [ 51%]
tests/test_piqs.py::TestDicke::test_liouvillian PASSED                   [ 51%]
tests/test_piqs.py::TestDicke::test_gamma1 PASSED                        [ 51%]
tests/test_piqs.py::TestDicke::test_gamma2 PASSED                        [ 51%]
tests/test_piqs.py::TestDicke::test_gamma3 PASSED                        [ 52%]
tests/test_piqs.py::TestDicke::test_gamma4 PASSED                        [ 52%]
tests/test_piqs.py::TestDicke::test_gamma5 PASSED                        [ 52%]
tests/test_piqs.py::TestDicke::test_gamma6 PASSED                        [ 52%]
tests/test_piqs.py::TestDicke::test_gamma7 PASSED                        [ 52%]
tests/test_piqs.py::TestDicke::test_gamma8 PASSED                        [ 52%]
tests/test_piqs.py::TestDicke::test_gamma9 PASSED                        [ 52%]
tests/test_piqs.py::TestPim::test_gamma PASSED                           [ 53%]
tests/test_piqs.py::TestPim::test_isdicke PASSED                         [ 53%]
tests/test_piqs.py::TestPim::test_isdiagonal PASSED                      [ 53%]
tests/test_piqs.py::TestPim::test_coefficient_matrix PASSED              [ 53%]
tests/test_piqs.py::TestPim::test_pisolve PASSED                         [ 53%]
tests/test_processor.py::TestCircuitProcessor::test_modify_ctrls PASSED  [ 53%]
tests/test_processor.py::TestCircuitProcessor::test_save_read PASSED     [ 54%]
tests/test_processor.py::TestCircuitProcessor::test_id_evolution PASSED  [ 54%]
tests/test_processor.py::TestCircuitProcessor::test_id_with_T1_T2 PASSED [ 54%]
tests/test_propagator.py::testPropHO PASSED                              [ 54%]
tests/test_propagator.py::testPropHOB PASSED                             [ 54%]
tests/test_propagator.py::testPropHOPar PASSED                           [ 54%]
tests/test_propagator.py::testPropHOStrTd PASSED                         [ 54%]
tests/test_propagator.py::testPropHOFuncTd PASSED                        [ 55%]
tests/test_propagator.py::testPropHOSteady PASSED                        [ 55%]
tests/test_propagator.py::testPropHOSteadyPar PASSED                     [ 55%]
tests/test_propagator.py::testPropHDims PASSED                           [ 55%]
tests/test_ptrace.py::test_ptrace_rand FAILED                            [ 55%]
tests/test_qft.py::TestQFT::testQFTComparison PASSED                     [ 55%]
tests/test_qft.py::TestQFT::testQFTGateSequenceNoSwapping PASSED         [ 56%]
tests/test_qft.py::TestQFT::testQFTGateSequenceWithSwapping PASSED       [ 56%]
tests/test_qobj.py::test_QobjData PASSED                                 [ 56%]
tests/test_qobj.py::test_QobjType PASSED                                 [ 56%]
tests/test_qobj.py::TestQobjHermicity::test_standard PASSED              [ 56%]
tests/test_qobj.py::TestQobjHermicity::test_addition PASSED              [ 56%]
tests/test_qobj.py::TestQobjHermicity::test_multiplication PASSED        [ 57%]
tests/test_qobj.py::test_QobjUnitaryOper PASSED                          [ 57%]
tests/test_qobj.py::test_QobjDimsShape PASSED                            [ 57%]
tests/test_qobj.py::test_QobjMulNonsquareDims PASSED                     [ 57%]
tests/test_qobj.py::test_QobjAddition PASSED                             [ 57%]
tests/test_qobj.py::test_QobjSubtraction PASSED                          [ 57%]
tests/test_qobj.py::test_QobjMultiplication PASSED                       [ 57%]
tests/test_qobj.py::test_QobjDivision PASSED                             [ 58%]
tests/test_qobj.py::test_QobjPower PASSED                                [ 58%]
tests/test_qobj.py::test_QobjNeg PASSED                                  [ 58%]
tests/test_qobj.py::test_QobjEquals PASSED                               [ 58%]
tests/test_qobj.py::test_QobjGetItem PASSED                              [ 58%]
tests/test_qobj.py::test_CheckMulType PASSED                             [ 58%]
tests/test_qobj.py::test_Qobj_Spmv PASSED                                [ 59%]
tests/test_qobj.py::test_QobjConjugate PASSED                            [ 59%]
tests/test_qobj.py::test_QobjDagger PASSED                               [ 59%]
tests/test_qobj.py::test_QobjDiagonals PASSED                            [ 59%]
tests/test_qobj.py::test_QobjEigenEnergies PASSED                        [ 59%]
tests/test_qobj.py::test_QobjEigenStates PASSED                          [ 59%]
tests/test_qobj.py::test_QobjExpm PASSED                                 [ 59%]
tests/test_qobj.py::test_QobjExpmExplicitlySparse PASSED                 [ 60%]
tests/test_qobj.py::test_QobjExpmZeroOper PASSED                         [ 60%]
tests/test_qobj.py::test_Qobj_sqrtm PASSED                               [ 60%]
tests/test_qobj.py::test_Qobj_inv PASSED                                 [ 60%]
tests/test_qobj.py::test_QobjFull PASSED                                 [ 60%]
tests/test_qobj.py::test_QobjNorm PASSED                                 [ 60%]
tests/test_qobj.py::test_QobjPurity PASSED                               [ 61%]
tests/test_qobj.py::test_QobjPermute PASSED                              [ 61%]
tests/test_qobj.py::test_KetType PASSED                                  [ 61%]
tests/test_qobj.py::test_BraType PASSED                                  [ 61%]
tests/test_qobj.py::test_OperType PASSED                                 [ 61%]
tests/test_qobj.py::test_SuperType PASSED                                [ 61%]
tests/test_qobj.py::test_dag_preserves_superrep[to_super-2] PASSED       [ 61%]
tests/test_qobj.py::test_dag_preserves_superrep[to_super-4] PASSED       [ 62%]
tests/test_qobj.py::test_dag_preserves_superrep[to_super-8] PASSED       [ 62%]
tests/test_qobj.py::test_dag_preserves_superrep[to_choi-2] PASSED        [ 62%]
tests/test_qobj.py::test_dag_preserves_superrep[to_choi-4] PASSED        [ 62%]
tests/test_qobj.py::test_dag_preserves_superrep[to_choi-8] PASSED        [ 62%]
tests/test_qobj.py::test_dag_preserves_superrep[to_chi-2] PASSED         [ 62%]
tests/test_qobj.py::test_dag_preserves_superrep[to_chi-4] PASSED         [ 63%]
tests/test_qobj.py::test_dag_preserves_superrep[to_chi-8] PASSED         [ 63%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[add-True-True-super] PASSED [ 63%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[add-True-True-choi] PASSED [ 63%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[add-True-True-chi] PASSED [ 63%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[sub-True-True-super] PASSED [ 63%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[sub-True-True-choi] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[sub-True-True-chi] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[mul-True-True-super] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[mul-True-True-choi] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[mul-True-True-chi] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[truediv-False-True-super] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[truediv-False-True-choi] PASSED [ 64%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[truediv-False-True-chi] PASSED [ 65%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[tensor-True-False-super] PASSED [ 65%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[tensor-True-False-choi] PASSED [ 65%]
tests/test_qobj.py::test_arithmetic_preserves_superrep[tensor-True-False-chi] PASSED [ 65%]
tests/test_qobj.py::test_isherm_skew PASSED                              [ 65%]
tests/test_qobj.py::test_super_tensor_operket PASSED                     [ 65%]
tests/test_qobj.py::test_super_tensor_property PASSED                    [ 66%]
tests/test_qobj.py::test_composite_oper PASSED                           [ 66%]
tests/test_qobj.py::test_composite_vec PASSED                            [ 66%]
tests/test_qobj.py::TestTruncNeg::test_positive_operator PASSED          [ 66%]
tests/test_qobj.py::TestTruncNeg::test_diagonal_operator PASSED          [ 66%]
tests/test_qobj.py::TestTruncNeg::test_nondiagonal_operator PASSED       [ 66%]
tests/test_qobj.py::TestTruncNeg::test_sgs_known_good PASSED             [ 66%]
tests/test_qobj.py::test_cosm PASSED                                     [ 67%]
tests/test_qobj.py::test_sinm PASSED                                     [ 67%]
tests/test_qobj.py::test_dual_channel[sub_dimensions0] PASSED            [ 67%]
tests/test_qobj.py::test_dual_channel[sub_dimensions1] PASSED            [ 67%]
tests/test_qobj.py::test_dual_channel[sub_dimensions2] PASSED            [ 67%]
tests/test_qobj.py::test_dual_channel[sub_dimensions3] PASSED            [ 67%]
tests/test_qobj.py::test_call PASSED                                     [ 68%]
tests/test_qobj.py::test_matelem PASSED                                  [ 68%]
tests/test_qobj.py::test_projection PASSED                               [ 68%]
tests/test_qobj.py::test_overlap PASSED                                  [ 68%]
tests/test_qobj.py::test_unit PASSED                                     [ 68%]
tests/test_qobjevo.py::test_QobjEvo_call PASSED                          [ 68%]
tests/test_qobjevo.py::test_QobjEvo_call_args PASSED                     [ 69%]
tests/test_qobjevo.py::test_QobjEvo_step_coeff PASSED                    [ 69%]
tests/test_qobjevo.py::test_QobjEvo_copy PASSED                          [ 69%]
tests/test_qobjevo.py::test_QobjEvo_to_list PASSED                       [ 69%]
tests/test_qobjevo.py::test_QobjEvo_math_arithmetic PASSED               [ 69%]
tests/test_qobjevo.py::test_QobjEvo_unitary PASSED                       [ 69%]
tests/test_qobjevo.py::test_QobjEvo_tidyup PASSED                        [ 69%]
tests/test_qobjevo.py::test_QobjEvo_compress PASSED                      [ 70%]
tests/test_qobjevo.py::test_QobjEvo_shift PASSED                         [ 70%]
tests/test_qobjevo.py::test_QobjEvo_apply PASSED                         [ 70%]
tests/test_qobjevo.py::test_QobjEvo_apply_decorator PASSED               [ 70%]
tests/test_qobjevo.py::test_QobjEvo_mul_vec PASSED                       [ 70%]
tests/test_qobjevo.py::test_QobjEvo_mul_mat PASSED                       [ 70%]
tests/test_qobjevo.py::test_QobjEvo_expect_psi PASSED                    [ 71%]
tests/test_qobjevo.py::test_QobjEvo_expect_rho PASSED                    [ 71%]
tests/test_qobjevo.py::test_QobjEvo_with_state PASSED                    [ 71%]
tests/test_qobjevo.py::test_QobjEvo_pickle PASSED                        [ 71%]
tests/test_qobjevo.py::test_QobjEvo_safepickle PASSED                    [ 71%]
tests/test_qobjevo.py::test_QobjEvo_superoperator PASSED                 [ 71%]
tests/test_qpt.py::test_qpt_snot PASSED                                  [ 71%]
tests/test_qpt.py::test_qpt_cnot PASSED                                  [ 72%]
tests/test_qubit_evolution.py::test_MESolverCase1 PASSED                 [ 72%]
tests/test_qubit_evolution.py::test_MESolverCase2 PASSED                 [ 72%]
tests/test_qubit_evolution.py::test_ESSolverCase1 PASSED                 [ 72%]
tests/test_qubit_evolution.py::test_MCSolverCase1 PASSED                 [ 72%]
tests/test_qubit_evolution.py::test_MCSolverCase2 PASSED                 [ 72%]
tests/test_qubitcircuit.py::TestQubitCircuit::testSWAPtoCNOT PASSED      [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testISWAPtoCNOT PASSED     [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testCSIGNtoCNOT PASSED     [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testCNOTtoCSIGN PASSED     [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testCNOTtoSQRTSWAP PASSED  [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testCNOTtoSQRTISWAP PASSED [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testCNOTtoISWAP PASSED     [ 73%]
tests/test_qubitcircuit.py::TestQubitCircuit::testSNOTdecompose PASSED   [ 74%]
tests/test_qubitcircuit.py::TestQubitCircuit::testadjacentgates PASSED   [ 74%]
tests/test_qubitcircuit.py::TestQubitCircuit::test_add_gate PASSED       [ 74%]
tests/test_qubitcircuit.py::TestQubitCircuit::test_add_state PASSED      [ 74%]
tests/test_qubitcircuit.py::TestQubitCircuit::test_reverse PASSED        [ 74%]
tests/test_qubitcircuit.py::TestQubitCircuit::test_user_gate PASSED      [ 74%]
tests/test_qubitcircuit.py::TestQubitCircuit::test_N_level_system PASSED [ 75%]
tests/test_qubits.py::TestQubits::testQubitStates PASSED                 [ 75%]
tests/test_rand.py::TestRand::testRandUnitary PASSED                     [ 75%]
tests/test_rand.py::TestRand::testRandUnitarySeed PASSED                 [ 75%]
tests/test_rand.py::TestRand::testRandherm PASSED                        [ 75%]
tests/test_rand.py::TestRand::testRandhermSeed PASSED                    [ 75%]
tests/test_rand.py::TestRand::testRandhermPosDef PASSED                  [ 76%]
tests/test_rand.py::TestRand::testRandhermEigs PASSED                    [ 76%]
tests/test_rand.py::TestRand::testRanddm PASSED                          [ 76%]
tests/test_rand.py::TestRand::testRandDmSeed PASSED                      [ 76%]
tests/test_rand.py::TestRand::testRanddmEigs PASSED                      [ 76%]
tests/test_rand.py::TestRand::testRandket PASSED                         [ 76%]
tests/test_rand.py::TestRand::testRandketSeed PASSED                     [ 76%]
tests/test_rand.py::TestRand::testRandStochasticLeft PASSED              [ 77%]
tests/test_rand.py::TestRand::testRandStochasticLeftSeed PASSED          [ 77%]
tests/test_rand.py::TestRand::testRandStochasticRight PASSED             [ 77%]
tests/test_rand.py::TestRand::testRandStochasticRightSeed PASSED         [ 77%]
tests/test_random.py::test_rand_unitary_haar_unitarity PASSED            [ 77%]
tests/test_random.py::test_rand_dm_ginibre_rank PASSED                   [ 77%]
tests/test_random.py::test_rand_super_bcsz_cptp PASSED                   [ 78%]
tests/test_random.py::test_rand_vector_dims PASSED                       [ 78%]
tests/test_random.py::test_rand_oper_dims PASSED                         [ 78%]
tests/test_random.py::test_rand_super_dims PASSED                        [ 78%]
tests/test_scattering.py::TestScattering::testScatteringProbability PASSED [ 78%]
tests/test_scattering.py::TestScattering::testScatteringAmplitude PASSED [ 78%]
tests/test_scattering.py::TestScattering::testWaveguideSplit PASSED      [ 78%]
tests/test_sesolve.py::TestSESolve::test_01_1_state_with_const_H PASSED  [ 79%]
tests/test_sesolve.py::TestSESolve::test_01_1_unitary_with_const_H PASSED [ 79%]
tests/test_sesolve.py::TestSESolve::test_02_1_state_with_func_H PASSED   [ 79%]
tests/test_sesolve.py::TestSESolve::test_02_2_unitary_with_func_H PASSED [ 79%]
tests/test_sesolve.py::TestSESolve::test_03_1_state_with_list_func_H PASSED [ 79%]
tests/test_sesolve.py::TestSESolve::test_03_2_unitary_with_list_func_H PASSED [ 79%]
tests/test_sesolve.py::TestSESolve::test_04_1_state_with_list_str_H PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_04_2_unitary_with_list_func_H PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_05_1_state_with_interp_H PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_05_2_unitary_with_interp_H PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_06_1_compare_state_and_unitary_const PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_06_2_compare_state_and_unitary_func PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_06_3_compare_state_and_unitary_list_func PASSED [ 80%]
tests/test_sesolve.py::TestSESolve::test_06_4_compare_state_and_unitary_list_str PASSED [ 81%]
tests/test_sesolve.py::TestSESolve::test_07_1_dynamic_args PASSED        [ 81%]
tests/test_sp_eigs.py::test_SparseHermValsVecs PASSED                    [ 81%]
tests/test_sp_eigs.py::test_SparseValsVecs PASSED                        [ 81%]
tests/test_sp_eigs.py::test_SparseValsOnly PASSED                        [ 81%]
tests/test_sp_eigs.py::test_DenseHermValsVecs PASSED                     [ 81%]
tests/test_sp_eigs.py::test_DenseValsVecs PASSED                         [ 82%]
tests/test_sp_eigs.py::test_DenseValsOnly PASSED                         [ 82%]
tests/test_sparse.py::test_sparse_symmetric_permute PASSED               [ 82%]
tests/test_sparse.py::test_sparse_nonsymmetric_permute PASSED            [ 82%]
tests/test_sparse.py::test_sparse_symmetric_reverse_permute PASSED       [ 82%]
tests/test_sparse.py::test_sparse_nonsymmetric_reverse_permute PASSED    [ 82%]
tests/test_sparse.py::test_sp_bandwidth PASSED                           [ 83%]
tests/test_sparse.py::test_sp_profile PASSED                             [ 83%]
tests/test_sparse.py::test_sp_one_norm PASSED                            [ 83%]
tests/test_sparse.py::test_sp_inf_norm PASSED                            [ 83%]
tests/test_spinchain.py::TestSpinChain::test_linear_ISWAP PASSED         [ 83%]
tests/test_spinchain.py::TestSpinChain::test_linear_SQRTISWAP PASSED     [ 83%]
tests/test_spinchain.py::TestSpinChain::test_linear_combination PASSED   [ 83%]
tests/test_spinchain.py::TestSpinChain::test_circular_ISWAP PASSED       [ 84%]
tests/test_spinchain.py::TestSpinChain::test_circular_SQRTISWAP PASSED   [ 84%]
tests/test_spinchain.py::TestSpinChain::test_circular_combination PASSED [ 84%]
tests/test_spinchain.py::TestSpinChain::test_analytical_evo PASSED       [ 84%]
tests/test_spinchain.py::TestSpinChain::test_numerical_evo PASSED        [ 84%]
tests/test_spmath.py::test_csr_kron PASSED                               [ 84%]
tests/test_spmath.py::test_zcsr_transpose PASSED                         [ 85%]
tests/test_spmath.py::test_zcsr_adjoint PASSED                           [ 85%]
tests/test_spmath.py::test_zcsr_mult PASSED                              [ 85%]
tests/test_spmath.py::test_zcsr_isherm PASSED                            [ 85%]
tests/test_states.py::test_basis_simple[2-0] PASSED                      [ 85%]
tests/test_states.py::test_basis_simple[2-1] PASSED                      [ 85%]
tests/test_states.py::test_basis_simple[100-99] PASSED                   [ 85%]
tests/test_states.py::test_implicit_tensor_basis_like[size0-n0-basis] PASSED [ 86%]
tests/test_states.py::test_implicit_tensor_basis_like[size0-n0-fock] PASSED [ 86%]
tests/test_states.py::test_implicit_tensor_basis_like[size0-n0-fock_dm] PASSED [ 86%]
tests/test_states.py::test_implicit_tensor_basis_like[size1-n1-basis] PASSED [ 86%]
tests/test_states.py::test_implicit_tensor_basis_like[size1-n1-fock] PASSED [ 86%]
tests/test_states.py::test_implicit_tensor_basis_like[size1-n1-fock_dm] PASSED [ 86%]
tests/test_states.py::test_implicit_tensor_projection[size0-n0-m0] PASSED [ 87%]
tests/test_states.py::test_implicit_tensor_projection[size1-n1-m1] PASSED [ 87%]
tests/test_states.py::TestStates::testCoherentState PASSED               [ 87%]
tests/test_states.py::TestStates::testCoherentDensityMatrix PASSED       [ 87%]
tests/test_states.py::TestStates::testThermalDensityMatrix PASSED        [ 87%]
tests/test_states.py::TestStates::testFockDensityMatrix PASSED           [ 87%]
tests/test_states.py::TestStates::testTripletStateNorm PASSED            [ 88%]
tests/test_steadystate.py::test_qubit_direct PASSED                      [ 88%]
tests/test_steadystate.py::test_qubit_eigen PASSED                       [ 88%]
tests/test_steadystate.py::test_qubit_power PASSED                       [ 88%]
tests/test_steadystate.py::test_qubit_power_gmres PASSED                 [ 88%]
tests/test_steadystate.py::test_qubit_power_bicgstab PASSED              [ 88%]
tests/test_steadystate.py::test_qubit_gmres PASSED                       [ 88%]
tests/test_steadystate.py::test_qubit_bicgstab PASSED                    [ 89%]
tests/test_steadystate.py::test_ho_direct PASSED                         [ 89%]
tests/test_steadystate.py::test_ho_eigen PASSED                          [ 89%]
tests/test_steadystate.py::test_ho_power PASSED                          [ 89%]
tests/test_steadystate.py::test_ho_power_gmres PASSED                    [ 89%]
tests/test_steadystate.py::test_ho_power_bicgstab PASSED                 [ 89%]
tests/test_steadystate.py::test_ho_gmres PASSED                          [ 90%]
tests/test_steadystate.py::test_ho_bicgstab PASSED                       [ 90%]
tests/test_steadystate.py::test_driven_cavity_direct PASSED              [ 90%]
tests/test_steadystate.py::test_driven_cavity_eigen PASSED               [ 90%]
tests/test_steadystate.py::test_driven_cavity_power PASSED               [ 90%]
tests/test_steadystate.py::test_driven_cavity_power_gmres PASSED         [ 90%]
tests/test_steadystate.py::test_driven_cavity_power_bicgstab PASSED      [ 90%]
tests/test_steadystate.py::test_driven_cavity_gmres PASSED               [ 91%]
tests/test_steadystate.py::test_driven_cavity_bicgstab PASSED            [ 91%]
tests/test_stochastic_me.py::test_smesolve_photocurrent FAILED           [ 91%]
tests/test_stochastic_me.py::test_smesolve_homodyne FAILED               [ 91%]
tests/test_stochastic_me.py::test_ssesolve_feedback FAILED               [ 91%]
tests/test_stochastic_se.py::test_ssesolve_photocurrent FAILED           [ 91%]
tests/test_stochastic_se.py::test_ssesolve_homodyne FAILED               [ 92%]
tests/test_stochastic_se.py::test_ssesolve_feedback FAILED               [ 92%]
tests/test_subsys_apply.py::TestSubsysApply::test_SimpleSingleApply PASSED [ 92%]
tests/test_subsys_apply.py::TestSubsysApply::test_SimpleSuperApply PASSED [ 92%]
tests/test_subsys_apply.py::TestSubsysApply::test_ComplexSingleApply PASSED [ 92%]
tests/test_subsys_apply.py::TestSubsysApply::test_ComplexSuperApply PASSED [ 92%]
tests/test_superop_reps.py::TestSuperopReps::test_SuperChoiSuper PASSED  [ 92%]
tests/test_superop_reps.py::TestSuperopReps::test_SuperChoiChiSuper PASSED [ 93%]
tests/test_superop_reps.py::TestSuperopReps::test_ChoiKrausChoi PASSED   [ 93%]
tests/test_superop_reps.py::TestSuperopReps::test_NonSquareKrausSuperChoi PASSED [ 93%]
tests/test_superop_reps.py::TestSuperopReps::test_NeglectSmallKraus PASSED [ 93%]
tests/test_superop_reps.py::TestSuperopReps::test_SuperPreservesSelf PASSED [ 93%]
tests/test_superop_reps.py::TestSuperopReps::test_ChoiPreservesSelf PASSED [ 93%]
tests/test_superop_reps.py::TestSuperopReps::test_random_iscptp PASSED   [ 94%]
tests/test_superop_reps.py::TestSuperopReps::test_known_iscptp PASSED    [ 94%]
tests/test_superop_reps.py::TestSuperopReps::test_choi_tr PASSED         [ 94%]
tests/test_superop_reps.py::TestSuperopReps::test_stinespring_cp PASSED  [ 94%]
tests/test_superop_reps.py::TestSuperopReps::test_stinespring_agrees PASSED [ 94%]
tests/test_superop_reps.py::TestSuperopReps::test_stinespring_dims PASSED [ 94%]
tests/test_superop_reps.py::TestSuperopReps::test_chi_choi_roundtrip PASSED [ 95%]
tests/test_superop_reps.py::TestSuperopReps::test_chi_known PASSED       [ 95%]
tests/test_superoper.py::TestMatVec::testOperatorVector PASSED           [ 95%]
tests/test_superoper.py::TestMatVec::testOperatorVectorTensor PASSED     [ 95%]
tests/test_superoper.py::TestMatVec::testOperatorVectorNotSquare PASSED  [ 95%]
tests/test_superoper.py::TestMatVec::testOperatorSpreAppl PASSED         [ 95%]
tests/test_superoper.py::TestMatVec::testOperatorSpostAppl PASSED        [ 95%]
tests/test_superoper.py::TestMatVec::testOperatorUnitaryTransform PASSED [ 96%]
tests/test_superoper.py::TestMatVec::testMatrixVecMat PASSED             [ 96%]
tests/test_superoper.py::TestMatVec::testVecMatVec PASSED                [ 96%]
tests/test_superoper.py::TestMatVec::testVecMatIndexConversion PASSED    [ 96%]
tests/test_superoper.py::TestMatVec::testVecMatIndexCompability PASSED   [ 96%]
tests/test_superoper.py::TestMatVec::test_reshuffle PASSED               [ 96%]
tests/test_superoper.py::TestMatVec::test_sprepost PASSED                [ 97%]
tests/test_superoper.py::TestMatVec::testLiouvillianImplem PASSED        [ 97%]
tests/test_td_formats.py::test_setTDFormatCheckMC PASSED                 [ 97%]
tests/test_tensor.py::test_tensor_contract_ident PASSED                  [ 97%]
tests/test_tensor.py::test_tensor_contract_other PASSED                  [ 97%]
tests/test_tensor.py::test_tensor_swap_other PASSED                      [ 97%]
tests/test_three_level.py::testThreeStates PASSED                        [ 97%]
tests/test_three_level.py::testThreeOps PASSED                           [ 98%]
tests/test_utilities.py::test_unit_conversions PASSED                    [ 98%]
tests/test_utilities.py::test_unit_clebsch PASSED                        [ 98%]
tests/test_wigner.py::test_wigner_bell1_su2parity PASSED                 [ 98%]
tests/test_wigner.py::test_parity PASSED                                 [ 98%]
tests/test_wigner.py::test_wigner_coherent PASSED                        [ 98%]
tests/test_wigner.py::test_wigner_fock PASSED                            [ 99%]
tests/test_wigner.py::test_wigner_compare_methods_dm PASSED              [ 99%]
tests/test_wigner.py::test_wigner_compare_methods_ket PASSED             [ 99%]
tests/test_wigner.py::test_wigner_fft_comparse_ket PASSED                [ 99%]
tests/test_wigner.py::test_wigner_fft_comparse_dm PASSED                 [ 99%]
tests/test_wigner.py::test_wigner_clenshaw_iter_dm PASSED                [ 99%]
tests/test_wigner.py::test_wigner_clenshaw_sp_iter_dm PASSED             [100%]

================================== FAILURES ===================================
_ test_transformations_from_qobj_and_direct_eigenbases_match[True-2-rand_bra] _

constructor = <function rand_bra at 0x000000000A56B048>, n_levels = 2
inverse = True

    @pytest.mark.parametrize("constructor", _state_constructors)
    @pytest.mark.parametrize("n_levels", [2, 10])
    @pytest.mark.parametrize("inverse", [True, False])
    def test_transformations_from_qobj_and_direct_eigenbases_match(constructor,
                                                                   n_levels,
                                                                   inverse):
        """
        Check transformations from the Qobj-calculated eigenbasis matches the one
        from the scipy-calculated basis.
        """
        qobj = constructor(n_levels)
        # Generate a random basis
        random_dm = qutip.rand_dm(n_levels, density=1)
        _, qobj_basis = random_dm.eigenstates()
        _, direct_basis = qutip.sparse.sp_eigs(random_dm.data, isherm=1)
>       H1 = qobj.transform(qobj_basis, inverse=inverse)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_basis_transformation.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\qobj.py:1457: in transform
    data = self.data.dot(S)
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:359: in dot
    return self * other
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:480: in __mul__
    return self._mul_sparse_matrix(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <1x2 sparse matrix of type '<class 'numpy.complex128'>'
	with 2 stored elements in Compressed Sparse Row format>
other = <2x2 sparse matrix of type '<class 'numpy.complex128'>'
	with 4 stored elements in Compressed Sparse Row format>

    def _mul_sparse_matrix(self, other):
        """
        Do the sparse matrix mult returning fast_csr_matrix only
        when other is also fast_csr_matrix.
        """
        M, K1 = self.shape
        K2, N = other.shape
    
        major_axis = self._swap((M,N))[0]
        if isinstance(other, fast_csr_matrix):
            A = zcsr_mult(self, other, sorted=1)
            return A
    
        other = csr_matrix(other)  # convert to this format
        idx_dtype = get_index_dtype((self.indptr, self.indices,
                                     other.indptr, other.indices),
                                    maxval=M*N)
        indptr = np.empty(major_axis + 1, dtype=idx_dtype)
    
>       fn = getattr(_sparsetools, self.format + '_matmat_pass1')
E       AttributeError: module 'scipy.sparse._sparsetools' has no attribute 'csr_matmat_pass1'

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\fastsparse.py:197: AttributeError
_ test_transformations_from_qobj_and_direct_eigenbases_match[True-10-rand_bra] _

constructor = <function rand_bra at 0x000000000A56B048>, n_levels = 10
inverse = True

    @pytest.mark.parametrize("constructor", _state_constructors)
    @pytest.mark.parametrize("n_levels", [2, 10])
    @pytest.mark.parametrize("inverse", [True, False])
    def test_transformations_from_qobj_and_direct_eigenbases_match(constructor,
                                                                   n_levels,
                                                                   inverse):
        """
        Check transformations from the Qobj-calculated eigenbasis matches the one
        from the scipy-calculated basis.
        """
        qobj = constructor(n_levels)
        # Generate a random basis
        random_dm = qutip.rand_dm(n_levels, density=1)
        _, qobj_basis = random_dm.eigenstates()
        _, direct_basis = qutip.sparse.sp_eigs(random_dm.data, isherm=1)
>       H1 = qobj.transform(qobj_basis, inverse=inverse)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_basis_transformation.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\qobj.py:1457: in transform
    data = self.data.dot(S)
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:359: in dot
    return self * other
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:480: in __mul__
    return self._mul_sparse_matrix(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <1x10 sparse matrix of type '<class 'numpy.complex128'>'
	with 10 stored elements in Compressed Sparse Row format>
other = <10x10 sparse matrix of type '<class 'numpy.complex128'>'
	with 100 stored elements in Compressed Sparse Row format>

    def _mul_sparse_matrix(self, other):
        """
        Do the sparse matrix mult returning fast_csr_matrix only
        when other is also fast_csr_matrix.
        """
        M, K1 = self.shape
        K2, N = other.shape
    
        major_axis = self._swap((M,N))[0]
        if isinstance(other, fast_csr_matrix):
            A = zcsr_mult(self, other, sorted=1)
            return A
    
        other = csr_matrix(other)  # convert to this format
        idx_dtype = get_index_dtype((self.indptr, self.indices,
                                     other.indptr, other.indices),
                                    maxval=M*N)
        indptr = np.empty(major_axis + 1, dtype=idx_dtype)
    
>       fn = getattr(_sparsetools, self.format + '_matmat_pass1')
E       AttributeError: module 'scipy.sparse._sparsetools' has no attribute 'csr_matmat_pass1'

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\fastsparse.py:197: AttributeError
_ test_transformations_from_qobj_and_direct_eigenbases_match[False-2-rand_bra] _

constructor = <function rand_bra at 0x000000000A56B048>, n_levels = 2
inverse = False

    @pytest.mark.parametrize("constructor", _state_constructors)
    @pytest.mark.parametrize("n_levels", [2, 10])
    @pytest.mark.parametrize("inverse", [True, False])
    def test_transformations_from_qobj_and_direct_eigenbases_match(constructor,
                                                                   n_levels,
                                                                   inverse):
        """
        Check transformations from the Qobj-calculated eigenbasis matches the one
        from the scipy-calculated basis.
        """
        qobj = constructor(n_levels)
        # Generate a random basis
        random_dm = qutip.rand_dm(n_levels, density=1)
        _, qobj_basis = random_dm.eigenstates()
        _, direct_basis = qutip.sparse.sp_eigs(random_dm.data, isherm=1)
>       H1 = qobj.transform(qobj_basis, inverse=inverse)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_basis_transformation.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\qobj.py:1467: in transform
    data = self.data.dot(S.conj().T)
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:359: in dot
    return self * other
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:480: in __mul__
    return self._mul_sparse_matrix(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <1x2 sparse matrix of type '<class 'numpy.complex128'>'
	with 2 stored elements in Compressed Sparse Row format>
other = <2x2 sparse matrix of type '<class 'numpy.complex128'>'
	with 2 stored elements in Compressed Sparse Row format>

    def _mul_sparse_matrix(self, other):
        """
        Do the sparse matrix mult returning fast_csr_matrix only
        when other is also fast_csr_matrix.
        """
        M, K1 = self.shape
        K2, N = other.shape
    
        major_axis = self._swap((M,N))[0]
        if isinstance(other, fast_csr_matrix):
            A = zcsr_mult(self, other, sorted=1)
            return A
    
        other = csr_matrix(other)  # convert to this format
        idx_dtype = get_index_dtype((self.indptr, self.indices,
                                     other.indptr, other.indices),
                                    maxval=M*N)
        indptr = np.empty(major_axis + 1, dtype=idx_dtype)
    
>       fn = getattr(_sparsetools, self.format + '_matmat_pass1')
E       AttributeError: module 'scipy.sparse._sparsetools' has no attribute 'csr_matmat_pass1'

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\fastsparse.py:197: AttributeError
_ test_transformations_from_qobj_and_direct_eigenbases_match[False-10-rand_bra] _

constructor = <function rand_bra at 0x000000000A56B048>, n_levels = 10
inverse = False

    @pytest.mark.parametrize("constructor", _state_constructors)
    @pytest.mark.parametrize("n_levels", [2, 10])
    @pytest.mark.parametrize("inverse", [True, False])
    def test_transformations_from_qobj_and_direct_eigenbases_match(constructor,
                                                                   n_levels,
                                                                   inverse):
        """
        Check transformations from the Qobj-calculated eigenbasis matches the one
        from the scipy-calculated basis.
        """
        qobj = constructor(n_levels)
        # Generate a random basis
        random_dm = qutip.rand_dm(n_levels, density=1)
        _, qobj_basis = random_dm.eigenstates()
        _, direct_basis = qutip.sparse.sp_eigs(random_dm.data, isherm=1)
>       H1 = qobj.transform(qobj_basis, inverse=inverse)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_basis_transformation.py:107: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\qobj.py:1467: in transform
    data = self.data.dot(S.conj().T)
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:359: in dot
    return self * other
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:480: in __mul__
    return self._mul_sparse_matrix(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <1x10 sparse matrix of type '<class 'numpy.complex128'>'
	with 10 stored elements in Compressed Sparse Row format>
other = <10x10 sparse matrix of type '<class 'numpy.complex128'>'
	with 100 stored elements in Compressed Sparse Row format>

    def _mul_sparse_matrix(self, other):
        """
        Do the sparse matrix mult returning fast_csr_matrix only
        when other is also fast_csr_matrix.
        """
        M, K1 = self.shape
        K2, N = other.shape
    
        major_axis = self._swap((M,N))[0]
        if isinstance(other, fast_csr_matrix):
            A = zcsr_mult(self, other, sorted=1)
            return A
    
        other = csr_matrix(other)  # convert to this format
        idx_dtype = get_index_dtype((self.indptr, self.indices,
                                     other.indptr, other.indices),
                                    maxval=M*N)
        indptr = np.empty(major_axis + 1, dtype=idx_dtype)
    
>       fn = getattr(_sparsetools, self.format + '_matmat_pass1')
E       AttributeError: module 'scipy.sparse._sparsetools' has no attribute 'csr_matmat_pass1'

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\fastsparse.py:197: AttributeError
______________________________ test_dqd_current _______________________________

    def test_dqd_current():
        "Counting statistics: current and current noise in a DQD model"
    
        G = 0
        L = 1
        R = 2
    
        sz = projection(3, L, L) - projection(3, R, R)
        sx = projection(3, L, R) + projection(3, R, L)
        sR = projection(3, G, R)
        sL = projection(3, G, L)
    
        w0 = 1
        tc = 0.6 * w0
        GammaR = 0.0075 * w0
        GammaL = 0.0075 * w0
        nth = 0.00
        eps_vec = np.linspace(-1.5*w0, 1.5*w0, 20)
    
        J_ops = [GammaR * sprepost(sR, sR.dag())]
    
        c_ops = [np.sqrt(GammaR * (1 + nth)) * sR,
                 np.sqrt(GammaR * (nth)) * sR.dag(),
                 np.sqrt(GammaL * (nth)) * sL,
                 np.sqrt(GammaL * (1 + nth)) * sL.dag(),
                 ]
    
        I = np.zeros(len(eps_vec))
        S = np.zeros(len(eps_vec))
    
        for n, eps in enumerate(eps_vec):
            H = (eps/2 * sz + tc * sx)
            L = liouvillian(H, c_ops)
            rhoss = steadystate(L)
>           I[n], S[n] = countstat_current_noise(L, [], rhoss=rhoss, J_ops=J_ops)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_countstat.py:74: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\countstat.py:248: in countstat_current_noise
    S[j, i, k] -= (expect_rho_vec(Jj.data * Q,
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:480: in __mul__
    return self._mul_sparse_matrix(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <9x9 sparse matrix of type '<class 'numpy.complex128'>'
	with 1 stored elements in Compressed Sparse Row format>
other = <9x9 sparse matrix of type '<class 'numpy.complex128'>'
	with 21 stored elements in Compressed Sparse Row format>

    def _mul_sparse_matrix(self, other):
        """
        Do the sparse matrix mult returning fast_csr_matrix only
        when other is also fast_csr_matrix.
        """
        M, K1 = self.shape
        K2, N = other.shape
    
        major_axis = self._swap((M,N))[0]
        if isinstance(other, fast_csr_matrix):
            A = zcsr_mult(self, other, sorted=1)
            return A
    
        other = csr_matrix(other)  # convert to this format
        idx_dtype = get_index_dtype((self.indptr, self.indices,
                                     other.indptr, other.indices),
                                    maxval=M*N)
        indptr = np.empty(major_axis + 1, dtype=idx_dtype)
    
>       fn = getattr(_sparsetools, self.format + '_matmat_pass1')
E       AttributeError: module 'scipy.sparse._sparsetools' has no attribute 'csr_matmat_pass1'

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\fastsparse.py:197: AttributeError
___________________________ test_fast_sparse_basic ____________________________

    def test_fast_sparse_basic():
        "fastsparse: fast_csr_matrix operations"
        H = rand_herm(5).data
        assert_(isinstance(H, fast_csr_matrix))
        Hadd = H + H
        assert_(isinstance(Hadd, fast_csr_matrix))
        Hsub = H + H
        assert_(isinstance(Hsub, fast_csr_matrix))
        Hmult = H * H
        assert_(isinstance(Hmult, fast_csr_matrix))
        Hcopy = H.copy()
        assert_(isinstance(Hcopy, fast_csr_matrix))
        assert_(isinstance(-1*H, fast_csr_matrix))
        assert(isinstance(H/2., fast_csr_matrix))
    
        G = sp.csr_matrix((H.data, H.indices,H.indptr),
                copy=True, shape=H.shape)
        assert_(not isinstance(G, fast_csr_matrix))
        Hadd = H + G
        assert_(not isinstance(Hadd, fast_csr_matrix))
        Hadd = G + H
        assert_(not isinstance(Hadd, fast_csr_matrix))
        Hsub = H - G
        assert_(not isinstance(Hsub, fast_csr_matrix))
        Hsub = G - H
        assert_(not isinstance(Hsub, fast_csr_matrix))
>       Hmult = H*G

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_fastsparse.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\base.py:480: in __mul__
    return self._mul_sparse_matrix(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <5x5 sparse matrix of type '<class 'numpy.complex128'>'
	with 20 stored elements in Compressed Sparse Row format>
other = <5x5 sparse matrix of type '<class 'numpy.complex128'>'
	with 20 stored elements in Compressed Sparse Row format>

    def _mul_sparse_matrix(self, other):
        """
        Do the sparse matrix mult returning fast_csr_matrix only
        when other is also fast_csr_matrix.
        """
        M, K1 = self.shape
        K2, N = other.shape
    
        major_axis = self._swap((M,N))[0]
        if isinstance(other, fast_csr_matrix):
            A = zcsr_mult(self, other, sorted=1)
            return A
    
        other = csr_matrix(other)  # convert to this format
        idx_dtype = get_index_dtype((self.indptr, self.indices,
                                     other.indptr, other.indices),
                                    maxval=M*N)
        indptr = np.empty(major_axis + 1, dtype=idx_dtype)
    
>       fn = getattr(_sparsetools, self.format + '_matmat_pass1')
E       AttributeError: module 'scipy.sparse._sparsetools' has no attribute 'csr_matmat_pass1'

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\fastsparse.py:197: AttributeError
____________________ TestLattice.test_bloch_wave_functions ____________________

self = <qutip.tests.test_lattice.TestLattice object at 0x000000000F587F88>

    def test_bloch_wave_functions(self):
        """
        lattice: Test the method Lattice1d.bloch_wave_functions().
        """
        # Coupled Resonator Optical Waveguide(CROW) Example(PhysRevB.99.224201)
        J = 2
        eta = np.pi/2
        H_cell = Qobj(np.array([[0, J*np.sin(eta)], [J*np.sin(eta), 0]]))
        inter_cell_T0 = (J/2)*Qobj(np.array([[np.exp(eta * 1j), 0], [0,
                                   np.exp(-eta*1j)]]))
        inter_cell_T1 = (J/2)*Qobj(np.array([[0, 1], [1, 0]]))
    
        inter_cell_T = [inter_cell_T0, inter_cell_T1]
    
        CROW_lattice = Lattice1d(num_cell=4, boundary="periodic",
                                 cell_num_site=1, cell_site_dof=[2],
                                 Hamiltonian_of_cell=H_cell,
                                 inter_hop=inter_cell_T)
        CROW_Haml = CROW_lattice.Hamiltonian()
    
        H_CROW = Qobj(np.array([[0.+0.j, 2.+0.j, 0.+1.j, 1.+0.j, 0.+0.j, 0.+0.j, 0.-1.j, 1.+0.j],
                                [2.+0.j, 0.+0.j, 1.+0.j, 0.-1.j, 0.+0.j, 0.+0.j, 1.+0.j, 0.+1.j],
                                [0.-1.j, 1.+0.j, 0.+0.j, 2.+0.j, 0.+1.j, 1.+0.j, 0.+0.j, 0.+0.j],
                                [1.+0.j, 0.+1.j, 2.+0.j, 0.+0.j, 1.+0.j, 0.-1.j, 0.+0.j, 0.+0.j],
                                [0.+0.j, 0.+0.j, 0.-1.j, 1.+0.j, 0.+0.j, 2.+0.j, 0.+1.j, 1.+0.j],
                                [0.+0.j, 0.+0.j, 1.+0.j, 0.+1.j, 2.+0.j, 0.+0.j, 1.+0.j, 0.-1.j],
                                [0.+1.j, 1.+0.j, 0.+0.j, 0.+0.j, 0.-1.j, 1.+0.j, 0.+0.j, 2.+0.j],
                                [1.+0.j, 0.-1.j, 0.+0.j, 0.+0.j, 1.+0.j, 0.+1.j, 2.+0.j, 0.+0.j]]),
                      dims=[[4, 2], [4, 2]])
        # Check for CROW with num_cell = 4
        assert_(np.max(abs(CROW_Haml-H_CROW)) < 1.0E-6)  # 1.0E-8 worked too
>       eigen_states = CROW_lattice.bloch_wave_functions()

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_lattice.py:437: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Lattice1d object: Number of cells = 4,
Number of sites in the cell = 1,
Degrees of freedom per site = [],
Lattice tens... 2), type = oper, isherm = True
Qobj data =
[[0. 1.]
 [1. 0.]]],
cell_tensor_config = [2]
Boundary Condition:  Periodic

    def bloch_wave_functions(self):
        """
        Returns eigenvectors ($\psi_n(k)$) of the Hamiltonian in a
        numpy.ndarray for translationally symmetric lattices with periodic
        boundary condition.
    
        .. math::
            :nowrap:
            \begin{eqnarray}
            |\psi_n(k) \rangle = |k \rangle \otimes | u_{n}(k) \rangle   \\
            | u_{n}(k) \rangle = a_n(k)|a\rangle  + b_n(k)|b\rangle \\
            \end{eqnarray}
    
        Please see section 1.2 of Asbóth, J. K., Oroszlány, L., & Pályi, A.
        (2016). A short course on topological insulators. Lecture notes in
        physics, 919 for a review.
    
        Returns
        -------
        eigenstates : ordered np.array
            eigenstates[j][0] is the jth eigenvalue.
            eigenstates[j][1] is the corresponding eigenvector.
        """
        if self.period_bnd_cond_x == 0:
            raise Exception("The lattice is not periodic.")
        (knxA, qH_ks, val_kns, vec_kns, vec_xs) = self._k_space_calculations()
        dtype = [('eigen_value', '<f16'), ('eigen_vector', Qobj)]
        values = list()
        for i in range(self.num_cell):
            for j in range(self._length_of_unit_cell):
                values.append((
                        val_kns[j][i], vec_xs[j+i*self._length_of_unit_cell]))
>       eigen_states = np.array(values, dtype=dtype)
E       TypeError: data type "<f16" not understood

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\lattice.py:860: TypeError
____________________________ TestLattice.test_CROW ____________________________

self = <qutip.tests.test_lattice.TestLattice object at 0x000000000FA55E48>

    def test_CROW(self):
        """
        lattice: Test the methods of Lattice1d in a CROW model.
        """
        # Coupled Resonator Optical Waveguide(CROW) Example(PhysRevB.99.224201)
        J = 2
        eta = np.pi/4
        H_cell = Qobj(np.array([[0, J * np.sin(eta)], [J * np.sin(eta), 0]]))
        inter_cell_T0 = (J/2)*Qobj(np.array([[np.exp(eta * 1j), 0],
                                   [0, np.exp(-eta*1j)]]))
        inter_cell_T1 = (J/2)*Qobj(np.array([[0, 1], [1, 0]]))
    
        inter_cell_T = [inter_cell_T0, inter_cell_T1]
    
        CROW_lattice = Lattice1d(num_cell=4, boundary="periodic",
                                 cell_num_site=1, cell_site_dof=[2],
                                 Hamiltonian_of_cell=H_cell,
                                 inter_hop=inter_cell_T)
        CROW_Haml = CROW_lattice.Hamiltonian()
    
        # Benchmark answers
        H_CROW = Qobj([[0.+0.j, 1.41421356+0.j, 0.70710678+0.70710678j, 1.+0.j,
                        0.+0.j, 0.+0.j, 0.70710678-0.70710678j, 1.+0.j],
        [1.41421356+0.j, 0.+0.j, 1.+0.j,  0.70710678-0.70710678j,
         0.+0.j, 0.+0.j, 1.+0.j, 0.70710678+0.70710678j],
        [0.70710678-0.70710678j, 1.+0.j, 0.+0.j, 1.41421356+0.j,
         0.70710678+0.70710678j, 1.+0.j, 0.+0.j, 0.+0.j],
        [1.+0.j, 0.70710678+0.70710678j, 1.41421356+0.j, 0.+0.j,
         1.+0.j, 0.70710678-0.70710678j, 0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j, 0.70710678-0.70710678j, 1.+0.j,
         0.+0.j, 1.41421356+0.j, 0.70710678+0.70710678j, 1.+0.j],
        [0.+0.j, 0.+0.j, 1.+0.j, 0.70710678+0.70710678j,
         1.41421356+0.j, 0.+0.j, 1.+0.j, 0.70710678-0.70710678j],
        [0.70710678+0.70710678j, 1.+0.j, 0.+0.j, 0.+0.j,
         0.70710678-0.70710678j, 1.+0.j, 0.+0.j, 1.41421356+0.j],
        [1.+0.j, 0.70710678-0.70710678j, 0.+0.j, 0.+0.j,
         1.+0.j, 0.70710678+0.70710678j, 1.41421356+0.j, 0.+0.j]],
         dims=[[4, 2], [4, 2]])
    
        # Check for CROW with num_cell = 4
        assert_(np.max(abs(CROW_Haml-H_CROW)) < 1.0E-6)  # 1.0E-8 worked too
    
        (kxA, val_kns) = CROW_lattice.get_dispersion()
        kCR = np.array([[-3.14159265],
                        [-1.57079633],
                        [0.],
                        [1.57079633]])
        vCR = np.array([[-2.        , -2.        , -2.        , -2.],
                        [-0.82842712, 2.         , 4.82842712 , 2.]])
        assert_(np.max(abs(kxA-kCR)) < 1.0E-6)
        assert_(np.max(abs(val_kns-vCR)) < 1.0E-6)
    
>       eigen_states = CROW_lattice.bloch_wave_functions()

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_lattice.py:502: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Lattice1d object: Number of cells = 4,
Number of sites in the cell = 1,
Degrees of freedom per site = [],
Lattice tens... 2), type = oper, isherm = True
Qobj data =
[[0. 1.]
 [1. 0.]]],
cell_tensor_config = [2]
Boundary Condition:  Periodic

    def bloch_wave_functions(self):
        """
        Returns eigenvectors ($\psi_n(k)$) of the Hamiltonian in a
        numpy.ndarray for translationally symmetric lattices with periodic
        boundary condition.
    
        .. math::
            :nowrap:
            \begin{eqnarray}
            |\psi_n(k) \rangle = |k \rangle \otimes | u_{n}(k) \rangle   \\
            | u_{n}(k) \rangle = a_n(k)|a\rangle  + b_n(k)|b\rangle \\
            \end{eqnarray}
    
        Please see section 1.2 of Asbóth, J. K., Oroszlány, L., & Pályi, A.
        (2016). A short course on topological insulators. Lecture notes in
        physics, 919 for a review.
    
        Returns
        -------
        eigenstates : ordered np.array
            eigenstates[j][0] is the jth eigenvalue.
            eigenstates[j][1] is the corresponding eigenvector.
        """
        if self.period_bnd_cond_x == 0:
            raise Exception("The lattice is not periodic.")
        (knxA, qH_ks, val_kns, vec_kns, vec_xs) = self._k_space_calculations()
        dtype = [('eigen_value', '<f16'), ('eigen_vector', Qobj)]
        values = list()
        for i in range(self.num_cell):
            for j in range(self._length_of_unit_cell):
                values.append((
                        val_kns[j][i], vec_xs[j+i*self._length_of_unit_cell]))
>       eigen_states = np.array(values, dtype=dtype)
E       TypeError: data type "<f16" not understood

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\lattice.py:860: TypeError
______________________________ test_ptrace_rand _______________________________

    def test_ptrace_rand():
        'ptrace : randomized tests, dense'
        for k in range(5):
            A = tensor(rand_ket(5), rand_ket(2), rand_ket(3))
            B = A.ptrace([1,2], False)
>           bdat,bd,bs = _pt(A, [1,2])

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_ptrace.py:114: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\legacy\ptrace.py:76: in _ptrace
    perm = perm.tocsr()
..\..\anaconda3\envs\qutip-env\lib\site-packages\scipy\sparse\lil.py:462: in tocsr
    _csparsetools.lil_get_lengths(self.rows, indptr[1:])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: Buffer has wrong number of dimensions (expected 1, got 2)

_csparsetools.pyx:109: ValueError
_________________________ test_smesolve_photocurrent __________________________

    def test_smesolve_photocurrent():
        "Stochastic: photocurrent_mesolve"
        tol = 0.01
    
        N = 4
        gamma = 0.25
        ntraj = 20
        nsubsteps = 100
        a = destroy(N)
    
        H = [[a.dag() * a,f]]
        psi0 = coherent(N, 0.5)
        sc_ops = [np.sqrt(gamma) * a, np.sqrt(gamma) * a * 0.5]
        e_ops = [a.dag() * a, a + a.dag(), (-1j)*(a - a.dag())]
    
        times = np.linspace(0, 1.0, 21)
        res_ref = mesolve(H, psi0, times, sc_ops, e_ops, args={"a":2})
        res = photocurrent_mesolve(H, psi0, times, [], sc_ops, e_ops, args={"a":2},
                       ntraj=ntraj, nsubsteps=nsubsteps,  store_measurement=True,
>                      map_func=parallel_map)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_stochastic_me.py:155: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:921: in photocurrent_mesolve
    args=args, **kwargs)
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:432: in __init__
    self.noise = np.random.randint(0, 2**32, ntraj).astype("u4")
mtrand.pyx:741: in numpy.random.mtrand.RandomState.randint
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: high is out of bounds for int32

_bounded_integers.pyx:1350: ValueError
___________________________ test_smesolve_homodyne ____________________________

    def test_smesolve_homodyne():
        "Stochastic: smesolve: homodyne, time-dependent H"
        tol = 0.01
    
        N = 4
        gamma = 0.25
        ntraj = 20
        nsubsteps = 100
        a = destroy(N)
    
        H = [[a.dag() * a,f]]
        psi0 = coherent(N, 0.5)
        sc_ops = [np.sqrt(gamma) * a, np.sqrt(gamma) * a * 0.5]
        e_ops = [a.dag() * a, a + a.dag(), (-1j)*(a - a.dag())]
    
        times = np.linspace(0, 1.0, 21)
        res_ref = mesolve(H, psi0, times, sc_ops, e_ops, args={"a":2})
        list_methods_tol = ['euler-maruyama',
                            'pc-euler',
                            'pc-euler-2',
                            'platen',
                            'milstein',
                            'milstein-imp',
                            'rouchon',
                            'taylor15',
                            'taylor15-imp',
                            'explicit15']
        for solver in list_methods_tol:
            res = smesolve(H, psi0, times, [], sc_ops, e_ops,
                           ntraj=ntraj, nsubsteps=nsubsteps, args={"a":2},
                           method='homodyne', store_measurement=True,
>                          solver=solver, map_func=parallel_map)

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_stochastic_me.py:195: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:589: in smesolve
    args=args, **kwargs)
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:432: in __init__
    self.noise = np.random.randint(0, 2**32, ntraj).astype("u4")
mtrand.pyx:741: in numpy.random.mtrand.RandomState.randint
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: high is out of bounds for int32

_bounded_integers.pyx:1350: ValueError
___________________________ test_ssesolve_feedback ____________________________

    def test_ssesolve_feedback():
        "Stochastic: ssesolve: time-dependent H with feedback"
        tol = 0.01
        N = 4
        ntraj = 10
        nsubsteps = 100
        a = destroy(N)
    
        H = [num(N)]
        psi0 = coherent(N, 2.5)
        sc_ops = [[a + a.dag(), f_dargs]]
        e_ops = [a.dag() * a, a + a.dag(), (-1j)*(a - a.dag()), qeye(N)]
    
        times = np.linspace(0, 10, 101)
        res_ref = mesolve(H, psi0, times, sc_ops, e_ops,
                          args={"expect_op_3":qeye(N)})
        res = smesolve(H, psi0, times, sc_ops=sc_ops, e_ops=e_ops, noise=1,
                       ntraj=ntraj, nsubsteps=nsubsteps, method='homodyne',
>                      map_func=parallel_map, args={"expect_op_3":qeye(N)})

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_stochastic_me.py:310: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:589: in smesolve
    args=args, **kwargs)
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:403: in __init__
    noise = np.random.randint(0, 2**32, ntraj)
mtrand.pyx:741: in numpy.random.mtrand.RandomState.randint
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: high is out of bounds for int32

_bounded_integers.pyx:1350: ValueError
_________________________ test_ssesolve_photocurrent __________________________

    def test_ssesolve_photocurrent():
        "Stochastic: photocurrent_sesolve"
        tol = 0.01
    
        N = 4
        gamma = 0.25
        ntraj = 25
        nsubsteps = 100
        a = destroy(N)
    
        H = [[a.dag() * a,f]]
        psi0 = coherent(N, 0.5)
        sc_ops = [np.sqrt(gamma) * a, np.sqrt(gamma) * a*0.5]
        e_ops = [a.dag() * a, a + a.dag(), (-1j)*(a - a.dag())]
    
        times = np.linspace(0, 2.5, 50)
        res_ref = mesolve(H, psi0, times, sc_ops, e_ops, args={"a":2})
        res = photocurrent_sesolve(H, psi0, times, sc_ops, e_ops, ntraj=ntraj,
                                  nsubsteps=nsubsteps, store_measurement=True,
>                                 map_func=parallel_map, args={"a":2})

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_stochastic_se.py:151: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:1009: in photocurrent_sesolve
    args=args, **kwargs)
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:432: in __init__
    self.noise = np.random.randint(0, 2**32, ntraj).astype("u4")
mtrand.pyx:741: in numpy.random.mtrand.RandomState.randint
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: high is out of bounds for int32

_bounded_integers.pyx:1350: ValueError
___________________________ test_ssesolve_homodyne ____________________________

    def test_ssesolve_homodyne():
        "Stochastic: ssesolve: homodyne, time-dependent H"
        tol = 0.01
    
        N = 4
        gamma = 0.25
        ntraj = 25
        nsubsteps = 100
        a = destroy(N)
    
        H = [[a.dag() * a,f]]
        psi0 = coherent(N, 0.5)
        sc_ops = [np.sqrt(gamma) * a, np.sqrt(gamma) * a*0.5]
        e_ops = [a.dag() * a, a + a.dag(), (-1j)*(a - a.dag())]
    
        times = np.linspace(0, 2.5, 50)
        res_ref = mesolve(H, psi0, times, sc_ops, e_ops, args={"a":2})
        res = ssesolve(H, psi0, times, sc_ops, e_ops,
                       ntraj=ntraj, nsubsteps=nsubsteps,
                       method='homodyne', store_measurement=True,
>                      map_func=parallel_map, args={"a":2})

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_stochastic_se.py:181: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:714: in ssesolve
    args=args, **kwargs)
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:432: in __init__
    self.noise = np.random.randint(0, 2**32, ntraj).astype("u4")
mtrand.pyx:741: in numpy.random.mtrand.RandomState.randint
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: high is out of bounds for int32

_bounded_integers.pyx:1350: ValueError
___________________________ test_ssesolve_feedback ____________________________

    def test_ssesolve_feedback():
        "Stochastic: ssesolve: time-dependent H with feedback"
        tol = 0.01
        N = 4
        ntraj = 10
        nsubsteps = 100
        a = destroy(N)
    
        H = [num(N)]
        psi0 = coherent(N, 2.5)
        sc_ops = [[a + a.dag(), f_dargs]]
        e_ops = [a.dag() * a, a + a.dag(), (-1j)*(a - a.dag()), qeye(N)]
    
        times = np.linspace(0, 10, 101)
        res_ref = mesolve(H, psi0, times, sc_ops, e_ops,
                          args={"expect_op_3":qeye(N)})
        res = ssesolve(H, psi0, times, sc_ops, e_ops, solver=None, noise=1,
                       ntraj=ntraj, nsubsteps=nsubsteps, method='homodyne',
>                      map_func=parallel_map, args={"expect_op_3":qeye(N)})

..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\tests\test_stochastic_se.py:244: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:714: in ssesolve
    args=args, **kwargs)
..\..\anaconda3\envs\qutip-env\lib\site-packages\qutip\stochastic.py:403: in __init__
    noise = np.random.randint(0, 2**32, ntraj)
mtrand.pyx:741: in numpy.random.mtrand.RandomState.randint
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: high is out of bounds for int32

_bounded_integers.pyx:1350: ValueError
=========================== short test summary info ===========================
FAILED tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-2-rand_bra]
FAILED tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[True-10-rand_bra]
FAILED tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-2-rand_bra]
FAILED tests/test_basis_transformation.py::test_transformations_from_qobj_and_direct_eigenbases_match[False-10-rand_bra]
FAILED tests/test_countstat.py::test_dqd_current - AttributeError: module 'sc...
FAILED tests/test_fastsparse.py::test_fast_sparse_basic - AttributeError: mod...
FAILED tests/test_lattice.py::TestLattice::test_bloch_wave_functions - TypeEr...
FAILED tests/test_lattice.py::TestLattice::test_CROW - TypeError: data type "...
FAILED tests/test_ptrace.py::test_ptrace_rand - ValueError: Buffer has wrong ...
FAILED tests/test_stochastic_me.py::test_smesolve_photocurrent - ValueError: ...
FAILED tests/test_stochastic_me.py::test_smesolve_homodyne - ValueError: high...
FAILED tests/test_stochastic_me.py::test_ssesolve_feedback - ValueError: high...
FAILED tests/test_stochastic_se.py::test_ssesolve_photocurrent - ValueError: ...
FAILED tests/test_stochastic_se.py::test_ssesolve_homodyne - ValueError: high...
FAILED tests/test_stochastic_se.py::test_ssesolve_feedback - ValueError: high...
= 15 failed, 616 passed, 9 skipped, 61 deselected, 2 xfailed, 959 warnings in 675.55s (0:11:15) =